velox/core/PlanNode.h

```diff
@@ -1782,4 +1782,52 @@ class WindowNode : public PlanNode {
   const RowTypePtr outputType_;
 };
 
+class MarkDistinctNode : public PlanNode {
```

Comment:
hashVariable actually is passed to Velox from Presto Coordinator.

Reply:
This sounds strange. Prestissimo runs with optimize_hash_generation = false and cannot run with optimize_hash_generation = true because Velox lacks the hash-generation function. Can you share the query plan?

Reply:
Also, I do not see hashVariables() being used anywhere. Let's remove it.

Reply:
Query plan
```

Fragment 3 [HASH]                                                                                                                                                                                                           
     CPU: 191.81ms, Scheduled: 223.40ms, Input: 6 rows (300B); per task: avg.: 0.06 std.dev.: 0.35, Output: 6 rows (258B)                                                                                                    
     Output layout: [field, c2, c3, c2$distinct, $hashvalue_45, $hashvalue_46]                                                                                                                                               
     Output partitioning: HASH [field, c3][$hashvalue_46]                                                                                                                                                                    
     Stage Execution Strategy: UNGROUPED_EXECUTION                                                                                                                                                                           
     - Project[projectLocality = LOCAL] => [field:integer, c2:bigint, c3:bigint, $hashvalue_45:bigint, $hashvalue_46:bigint, c2$distinct:boolean]                                                                            
             CPU: 0.00ns (0.00%), Scheduled: 4.00ms (12.90%), Output: 6 rows (258B)                                                                                                                                          
             Input avg.: 0.00 rows, Input std.dev.: 2272.30%                                                                                                                                                                 
         - MarkDistinct[distinct=field:integer, c2:bigint marker=c2$distinct][$hashvalue_47] => [field:integer, c2:bigint, c3:bigint, $hashvalue_45:bigint, $hashvalue_46:bigint, $hashvalue_47:bigint, c2$distinct:boolean] 
                 CPU: 0.00ns (0.00%), Scheduled: 0.00ns (0.00%), Output: 6 rows (312B)                                                                                                                                       
                 Input avg.: 0.00 rows, Input std.dev.: 2272.30%                                                                                                                                                             
             - LocalExchange[HASH][$hashvalue_47] (field, c2) => [field:integer, c2:bigint, c3:bigint, $hashvalue_45:bigint, $hashvalue_46:bigint, $hashvalue_47:bigint]                                                     
                     CPU: 0.00ns (0.00%), Scheduled: 0.00ns (0.00%), Output: 6 rows (300B)                                                                                                                                   
                     Input avg.: 0.00 rows, Input std.dev.: 2272.30%                                                                                                                                                         
                 - RemoteSource[4] => [field:integer, c2:bigint, c3:bigint, $hashvalue_48:bigint, $hashvalue_49:bigint, $hashvalue_50:bigint]                                                                                
                         CPU: 0.00ns (0.00%), Scheduled: 0.00ns (0.00%), Output: 6 rows (300B)                                                                                                                               
                         Input avg.: 0.00 rows, Input std.dev.: 2272.30%                   
```

Reply:
@shiyu-bytedance Would you share full plan? I'm curious how $hashvalue_47 is calculated. Also, just to confirm, this is a query plan from a query that runs on Prestissimo, not Presto Java.

Reply:
Full plan.

```
presto> explain analyze WITH sample AS ( SELECT * FROM (VALUES (1,1,1), (1,1,2), (1,1,1), (2,1,2), (2,2,1), (2,3,2)) AS account (c1, c2, c3) ) SELECT c1,sum(distinct c2),sum(distinct c3) from sample group by c1;                                                                           Query Plan                                                                            
-----------------------------------------------------------------------------------------------------------------------------------------------------------------
 Fragment 1 [HASH]                                                                                                                                               
     CPU: 7.11ms, Scheduled: 7.53ms, Input: 4 rows (416B); per task: avg.: 1.33 std.dev.: 0.94, Output: 2 rows (59.50kB)                                         
     Output layout: [field, sum, sum_25]                                                                                                                         
     Output partitioning: SINGLE []                                                                                                                              
     Stage Execution Strategy: UNGROUPED_EXECUTION                                                                                                               
     - Aggregate(FINAL)[field] => [field:integer, sum:bigint, sum_25:bigint]                                                                                     
             CPU: 0.00ns (0.00%), Scheduled: 0.00ns (0.00%), Output: 2 rows (59.50kB)                                                                            
             Input avg.: 0.08 rows, Input std.dev.: 479.58%                                                                                                      
             sum := "presto.default.sum"((sum_38)) (1:146)                                                                                                       
             sum_25 := "presto.default.sum"((sum_39)) (1:163)                                                                                                    
         - LocalExchange[HASH] (field) => [field:integer, sum_38:bigint, sum_39:bigint]                                                                          
                 CPU: 0.00ns (0.00%), Scheduled: 0.00ns (0.00%), Output: 4 rows (416B)                                                                           
                 Input avg.: 0.08 rows, Input std.dev.: 479.58%                                                                                                  
             - RemoteSource[2] => [field:integer, sum_38:bigint, sum_39:bigint]                                                                                  
                     CPU: 0.00ns (0.00%), Scheduled: 0.00ns (0.00%), Output: 4 rows (416B)                                                                       
                     Input avg.: 0.08 rows, Input std.dev.: 479.58%                                                                                              
                                                                                                                                                                 
 Fragment 2 [HASH]                                                                                                                                               
     CPU: 8.98ms, Scheduled: 9.12ms, Input: 6 rows (640B); per task: avg.: 2.00 std.dev.: 0.82, Output: 4 rows (119.03kB)                                        
     Output layout: [field, sum_38, sum_39]                                                                                                                      
     Output partitioning: HASH [field]                                                                                                                           
     Stage Execution Strategy: UNGROUPED_EXECUTION                                                                                                               
     - Aggregate(PARTIAL)[field] => [field:integer, sum_38:bigint, sum_39:bigint]                                                                                
             CPU: 0.00ns (0.00%), Scheduled: 0.00ns (0.00%), Output: 4 rows (119.03kB)                                                                           
             Input avg.: 0.13 rows, Input std.dev.: 420.32%                                                                                                      
             sum_38 := "presto.default.sum"((c2)) (mask = c2$distinct) (1:146)                                                                                   
             sum_39 := "presto.default.sum"((c3)) (mask = c3$distinct) (1:163)                                                                                   
         - MarkDistinct[distinct=field:integer, c3:bigint marker=c3$distinct] => [field:integer, c2:bigint, c3:bigint, c2$distinct:boolean, c3$distinct:boolean] 
                 CPU: 0.00ns (0.00%), Scheduled: 0.00ns (0.00%), Output: 6 rows (800B)                                                                           
                 Input avg.: 0.13 rows, Input std.dev.: 420.32%                                                                                                  
             - LocalExchange[HASH] (field, c3) => [field:integer, c2:bigint, c3:bigint, c2$distinct:boolean]                                                     
                     CPU: 0.00ns (0.00%), Scheduled: 0.00ns (0.00%), Output: 6 rows (640B)                                                                       
                     Input avg.: 0.13 rows, Input std.dev.: 420.32%                                                                                              
                 - RemoteSource[3] => [field:integer, c2:bigint, c3:bigint, c2$distinct:boolean]                                                                 
                         CPU: 0.00ns (0.00%), Scheduled: 0.00ns (0.00%), Output: 6 rows (640B)                                                                   
                         Input avg.: 0.13 rows, Input std.dev.: 420.32%                                                                                          
                                                                                                                                                                 
 Fragment 3 [HASH]                                                                                                                                               
     CPU: 8.27ms, Scheduled: 8.65ms, Input: 6 rows (576B); per task: avg.: 2.00 std.dev.: 2.16, Output: 6 rows (1.31kB)                                          
     Output layout: [field, c2, c3, c2$distinct]                                                                                                                 
     Output partitioning: HASH [field, c3]                                                                                                                       
     Stage Execution Strategy: UNGROUPED_EXECUTION                                                                                                               
     - MarkDistinct[distinct=field:integer, c2:bigint marker=c2$distinct] => [field:integer, c2:bigint, c3:bigint, c2$distinct:boolean]                          
             CPU: 0.00ns (0.00%), Scheduled: 0.00ns (0.00%), Output: 6 rows (768B)                                                                               
             Input avg.: 0.13 rows, Input std.dev.: 580.23%                                                                                                      
         - LocalExchange[HASH] (field, c2) => [field:integer, c2:bigint, c3:bigint]                                                                              
                 CPU: 0.00ns (0.00%), Scheduled: 0.00ns (0.00%), Output: 6 rows (576B)                                                                           
                 Input avg.: 0.13 rows, Input std.dev.: 580.23%                                                                                                  
             - RemoteSource[4] => [field:integer, c2:bigint, c3:bigint]                                                                                          
                     CPU: 0.00ns (0.00%), Scheduled: 0.00ns (0.00%), Output: 6 rows (576B)                                                                       
                     Input avg.: 0.13 rows, Input std.dev.: 580.23%                                                                                              
                                                                                                                                                                 
 Fragment 4 [SINGLE]                                                                                                                                             
     CPU: 1.34ms, Scheduled: 1.40ms, Input: 0 rows (0B); per task: avg.: 0.00 std.dev.: 0.00, Output: 6 rows (480B)                                              
     Output layout: [field, c2, c3]                                                                                                                              
     Output partitioning: HASH [field, c2]                                                                                                                       
     Stage Execution Strategy: UNGROUPED_EXECUTION                                                                                                               
     - Project[projectLocality = LOCAL] => [field:integer, c2:bigint, c3:bigint]                                                                                 
             CPU: 0.00ns (0.00%), Scheduled: 0.00ns (0.00%), Output: 6 rows (126B)                                                                               
             Input avg.: 0.38 rows, Input std.dev.: 387.30%                                                                                                      
             c2 := CAST(field_0 AS bigint) (1:35)                                                                                                                
             c3 := CAST(field_1 AS bigint) (1:35)                                                                                                                
         - LocalExchange[ROUND_ROBIN] () => [field:integer, field_0:integer, field_1:integer]                                                                    
                 CPU: 0.00ns (0.00%), Scheduled: 0.00ns (0.00%), Output: 6 rows (90B)                                                                            
                 Input avg.: 6.00 rows, Input std.dev.: 0.00%                                                                                                    
             - Values => [field:integer, field_0:integer, field_1:integer]                                                                                       
                     CPU: 0.00ns (0.00%), Scheduled: 0.00ns (0.00%), Output: 6 rows (96B)                                                                        
                     Input avg.: 0.00 rows, Input std.dev.: ?%                                                                                                   
                     (INTEGER'1', INTEGER'1', INTEGER'1')                                                                                                        
                     (INTEGER'1', INTEGER'1', INTEGER'2')                                                                                                        
                     (INTEGER'1', INTEGER'1', INTEGER'1')                                                                                                        
                     (INTEGER'2', INTEGER'1', INTEGER'2')                                                                                                        
                     (INTEGER'2', INTEGER'2', INTEGER'1')                                                                                                        
                     (INTEGER'2', INTEGER'3', INTEGER'2')                                                                                                        
                                                                                                                                                                 
                                                                                                                                                                 
(1 row)
```
This is coordinator + prestissimo worker

Reply:
@shiyu-bytedance Thank you for sharing the full plan. As expected, there are no $hashvalue_N columns here and  MarkDistinct plan nodes do not have 'hashVariable'.